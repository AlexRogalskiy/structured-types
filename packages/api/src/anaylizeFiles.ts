import * as ts from 'typescript';
import { dirname } from 'path';
import { PropTypes, PropType, PropDiagnostic, FunctionProp } from './types';
import { tsDefaults, DocsOptions, ProgramOptions } from './ts-utils';
import { SymbolParser } from './SymbolParser';

const cleanPropParemts = (
  prop: FunctionProp,
  propsName: 'properties' | 'parameters' | 'types',
  parents: Record<string, PropType>,
) => {
  if (prop[propsName]) {
    if (prop.parent && parents[prop.parent]) {
      delete prop[propsName];
    } else {
      prop[propsName] = consolidateParentProps(
        prop[propsName] as PropType[],
        parents,
      );
    }
  }
};
const consolidateParentProps = (
  props: PropType[],
  parents: Record<string, PropType>,
): PropType[] => {
  return props.map((prop) => {
    const propFn = prop as FunctionProp;
    cleanPropParemts(propFn, 'properties', parents);
    cleanPropParemts(propFn, 'parameters', parents);
    cleanPropParemts(propFn, 'types', parents);
    return propFn;
  });
};

export const anaylizeFiles = (
  fileNames: string[],
  options: DocsOptions = {},
  programOptions: ProgramOptions = {},
): PropTypes => {
  const {
    tsOptions: userTsOptions,
    scope = 'exports',
    ...parseOptions
  } = options;
  const tsOptions = { ...tsDefaults, ...userTsOptions };
  const {
    extractNames,
    collectDiagnostics,
    internalTypes,
    consolidateParents,
  } = parseOptions || {};
  const { program: userProgram, host: userHost } = programOptions;
  const host = userHost || ts.createCompilerHost(tsOptions);
  const program = userProgram || ts.createProgram(fileNames, tsOptions, host);
  const defaultLibraryFileName = host.getDefaultLibFileName(tsOptions);
  const defaultLibraryPath = host.getDefaultLibLocation
    ? host.getDefaultLibLocation()
    : dirname(defaultLibraryFileName);
  const equalityComparer = host.useCaseSensitiveFileNames()
    ? (s1: string, s2: string) => s1 === s2
    : (s1: string, s2: string) => s1.toUpperCase() === s2.toUpperCase();

  const isLibraryFile =
    defaultLibraryPath === '/'
      ? (file: ts.SourceFile) => {
          if (file.hasNoDefaultLib) {
            return true;
          }
          return equalityComparer(file.fileName, defaultLibraryFileName);
        }
      : (file: ts.SourceFile) => {
          if (file.hasNoDefaultLib) {
            return true;
          }
          return equalityComparer(dirname(file.fileName), defaultLibraryPath);
        };
  const parser = new SymbolParser(
    program,
    isLibraryFile,
    fileNames,
    parseOptions,
  );
  let parsed: PropTypes = {};
  const addSymbol = (symbol?: ts.Symbol): void => {
    if (symbol) {
      const symbolName = symbol.getName();
      if (
        internalTypes?.[symbolName] === undefined &&
        (!extractNames || extractNames.includes(symbolName))
      ) {
        const prop = parser.parse(symbol);
        if (prop) {
          parsed[symbolName] = prop;
        }
      }
    }
  };
  for (const fileName of fileNames) {
    const sourceFile = program.getSourceFile(fileName);
    if (sourceFile) {
      if (scope === 'all') {
        ts.forEachChild(sourceFile, (node: ts.Node) => {
          const namedNode = node as ts.ClassDeclaration;
          if (namedNode.name) {
            const symbol = parser.checker.getSymbolAtLocation(namedNode.name);
            addSymbol(symbol);
          } else if (ts.isVariableStatement(node)) {
            node.declarationList.declarations.forEach((d) => {
              if (d.name) {
                const symbol = parser.checker.getSymbolAtLocation(d.name);
                addSymbol(symbol);
              }
            });
          }
        });
      }
      const module = parser.checker.getSymbolAtLocation(sourceFile);
      if (module) {
        const exports = parser.checker.getExportsOfModule(module);
        exports.forEach((symbol) => addSymbol(symbol));
      }
    }
  }
  if (consolidateParents) {
    // only return parents that are not already exported from the same file
    const parents: Record<string, PropType> = Object.keys(parser.parents)
      .filter((name) => parsed[name] === undefined)
      .reduce((acc, name) => ({ ...acc, [name]: parser.parents[name] }), {});
    if (Object.keys(parents).length) {
      parsed = Object.keys(parsed).reduce((acc, key) => {
        return {
          ...acc,
          [key]: consolidateParentProps([parsed[key]], parents)[0],
        };
      }, {});
      parsed.__parents = Object.keys(parents).reduce((acc, key) => {
        return {
          ...acc,
          [key]: consolidateParentProps([parents[key]], parents)[0],
        };
      }, {});
    }
  }
  if (collectDiagnostics) {
    const allDiagnostics = ts
      .getPreEmitDiagnostics(program)
      .filter(({ file }) => file && fileNames.includes(file.fileName));
    if (allDiagnostics.length) {
      parsed.__diagnostics = allDiagnostics.map(
        ({ category, messageText, file, start }) => {
          const message =
            typeof messageText === 'string'
              ? messageText
              : messageText.messageText;

          const result: PropDiagnostic = {
            category,
            message,
          };
          if (file) {
            result.fileName = file.fileName.split('\\').pop()?.split('/').pop();
            if (typeof start !== 'undefined') {
              const location = file.getLineAndCharacterOfPosition(start);
              result.row = location.line + 1;
              result.column = location.character + 1;
            }
          }
          return result;
        },
      );
    }
  }

  return parsed;
};
