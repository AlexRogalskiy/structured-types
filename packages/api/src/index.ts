import { getTypescriptConfig } from '@structured-types/typescript-config';
import { analyzeFiles } from './analyzeFiles';
import { PropTypes } from './types';
import { tsDefaults, DocsOptions, ProgramOptions } from './ts-utils';
export * from './jsdoc';
export * from './ts';
export * from './types';
export * from './ts-utils';

export { analyzeFiles };

/**
 * API to analyze the given files by also loading the local typescript options from tsconfig
 * @param files list of files to be processed
 * @param options parsing options
 * @param programOptions typescript ts.program and ts.compilerHost
 * @returns the parsed types
 */
export const parseFiles = (
  files: string[],
  options: DocsOptions = {},
  programOptions?: ProgramOptions,
): PropTypes => {
  if (!files.length) {
    throw new Error('You need to supply at least one file');
  }
  options.tsOptions = {
    ...tsDefaults,
    ...getTypescriptConfig(files[0], options.tsOptions),
  };
  const results = analyzeFiles(files, options, programOptions);
  return results;
};
