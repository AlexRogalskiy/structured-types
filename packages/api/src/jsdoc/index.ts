export * from './mergeJSDoc';
export * from './parseJSDocTags';
