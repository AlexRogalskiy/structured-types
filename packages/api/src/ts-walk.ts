import * as ts from 'typescript';
import { PropTypes, PropDiagnostic } from './types';
import { tsDefaults, DocsOptions, ProgramOptions } from './ts-utils';
import { SymbolParser } from './SymbolParser';

function isNodeExported(node: ts.Node): boolean {
  return (
    (ts.getCombinedModifierFlags(node as ts.Declaration) &
      ts.ModifierFlags.Export) !==
      0 ||
    (!!node.parent && node.parent.kind === ts.SyntaxKind.SourceFile)
  );
}

export const anaylizeFiles = (
  fileNames: string[],
  options: DocsOptions = {},
  programOptions: ProgramOptions = {},
): PropTypes => {
  const {
    tsOptions = tsDefaults,
    scope = 'exports',
    ...parseOptions
  } = options;
  const { extractNames, collectDiagnostics } = parseOptions || {};
  const { program: userProgram, host } = programOptions;
  const program = userProgram || ts.createProgram(fileNames, tsOptions, host);
  // Get the checker, we will use it to find more about classes
  const checker = program.getTypeChecker();

  const parser = new SymbolParser(checker, parseOptions);
  const parsed: PropTypes = {};
  for (const fileName of fileNames) {
    const sourceFile = program.getSourceFile(fileName);
    if (sourceFile) {
      ts.forEachChild(sourceFile, (node: ts.Node) => {
        if (!ts.isModuleDeclaration(node)) {
          const include = scope === 'all' || isNodeExported(node);
          if (include) {
            const namedNode = node as ts.ClassDeclaration;
            if (namedNode.name) {
              const symbol = checker.getSymbolAtLocation(namedNode.name);
              if (symbol) {
                const symbolName = symbol.getName();
                if (!extractNames || extractNames.includes(symbolName)) {
                  const prop = parser.parseSymbol(symbol);
                  if (prop) {
                    parsed[symbolName] = prop;
                  }
                }
              }
            }
          }
        }
      });
    }
  }
  // only return parents that are not already exported from the same file
  const parents = Object.keys(parser.parents)
    .filter((name) => parsed[name] === undefined)
    .reduce((acc, name) => ({ ...acc, [name]: parser.parents[name] }), {});
  if (Object.keys(parents).length) {
    parsed.__parents = parents;
  }
  if (collectDiagnostics) {
    const allDiagnostics = ts
      .getPreEmitDiagnostics(program)
      .filter(({ file }) => file && fileNames.includes(file.fileName));
    if (allDiagnostics.length) {
      parsed.__diagnostics = allDiagnostics.map(
        ({ category, messageText, file, start }) => {
          const message =
            typeof messageText === 'string'
              ? messageText
              : messageText.messageText;

          const result: PropDiagnostic = {
            category,
            message,
          };
          if (file) {
            result.fileName = file.fileName.split('\\').pop()?.split('/').pop();
            if (typeof start !== 'undefined') {
              const location = file.getLineAndCharacterOfPosition(start);
              result.row = location.line + 1;
              result.column = location.character + 1;
            }
          }
          return result;
        },
      );
    }
  }

  return parsed;
};
