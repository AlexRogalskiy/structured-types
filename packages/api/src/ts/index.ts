export * from './resolveType';
export * from './getInitializer';
