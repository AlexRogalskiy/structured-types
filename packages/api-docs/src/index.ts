import * as path from 'path';
import { PropTypes } from '@structured-types/api';
import { cosmiconfigSync } from 'cosmiconfig';
import { CosmiconfigResult } from 'cosmiconfig/dist/types';

import { DocumentationNode, DocumentationOptions } from './types';
import { PropsToDocumentation } from './PropsToDocumentation';
export * from './types';

/**
 * Creates a list of api documentation nodes
 * @param props properties parsed from `structured-types/api`
 * @param options page generation options
 * @returns a list of documentation nodes
 */
export const propsToDocumentation = (
  props: PropTypes,
  options?: DocumentationOptions,
): DocumentationNode[] => {
  const extractor = new PropsToDocumentation();
  return extractor.extract(props, options);
};

/**
 * Read the api-docs configuration file
 * @param fileName the file that is being analyzed, will be used the starting folder to search for configuration files.
 * @param configFileName pass directly the configuration file name
 * @returns page generation options from the config file
 */
export const apiDocsConfig = (
  fileName: string,
  configFileName?: string,
): CosmiconfigResult => {
  const configExplorer = cosmiconfigSync('api-docs');
  if (configFileName) {
    return configExplorer.load(configFileName);
  }
  const searchPath = path.dirname(fileName);
  return configExplorer.search(searchPath);
};

export { getRepoPath } from './package-info/package-info';
