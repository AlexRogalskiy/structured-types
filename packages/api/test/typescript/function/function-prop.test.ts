import path from 'path';
import { parseFiles } from '../../../src/index';

describe('function', () => {
  it('extends-parameter', () => {
    const results = parseFiles(
      [path.resolve(__dirname, 'extends-parameter.ts')],
      { collectHelpers: true },
    );
    expect(results).toEqual({
      paintHomeyBear: {
        name: 'paintHomeyBear',
        kind: 11,
        parameters: [
          {
            kind: 15,
            type: 'ExtendT',
            properties: [
              {
                name: 'm',
                parent: 'T',
                kind: 1,
                description: 'base type member property',
              },
              {
                name: 'honey',
                kind: 3,
                description: 'own member',
              },
            ],
            name: 'ExtendT',
            description: 'extended type',
          },
        ],
        returns: {
          kind: 14,
          type: 'Bear',
          extends: ['Internal'],
          properties: [
            {
              name: 'honey',
              kind: 3,
              description: 'boolean type member',
            },
            {
              name: 'm',
              parent: 'Internal',
              kind: 1,
              description: 'string type member',
            },
          ],
          name: 'Bear',
          description: 'interface extending another one',
        },
        description: 'exported function',
      },
      __helpers: {
        Bear: {
          description: 'interface extending another one',
          kind: 14,
          name: 'Bear',
          extends: ['Internal'],
          properties: [
            {
              description: 'boolean type member',
              kind: 3,
              name: 'honey',
            },
            {
              description: 'string type member',
              kind: 1,
              name: 'm',
              parent: 'Internal',
            },
          ],
        },
        ExtendT: {
          description: 'extended type',
          kind: 15,
          name: 'ExtendT',
          properties: [
            {
              description: 'base type member property',
              kind: 1,
              name: 'm',
              parent: 'T',
            },
            {
              description: 'own member',
              kind: 3,
              name: 'honey',
            },
          ],
        },

        T: {
          name: 'T',
          kind: 15,
          properties: [
            {
              name: 'm',
              kind: 1,
              description: 'base type member property',
            },
          ],
          description: 'base type',
        },

        Internal: {
          name: 'Internal',
          kind: 14,
          properties: [
            {
              name: 'm',
              kind: 1,
              description: 'string type member',
            },
          ],
          description: 'internal interface with one member',
        },
      },
    });
  });
  it('inline-props', () => {
    const results = parseFiles([path.resolve(__dirname, 'inline-props.ts')], {
      collectHelpers: true,
    });

    expect(results).toEqual({
      inlineProps: {
        name: 'inlineProps',
        kind: 11,
        parameters: [
          {
            name: 'props',
            kind: 15,
            properties: [
              {
                name: 'stringProp',
                kind: 1,
              },
              {
                name: 'numProp',
                kind: 2,
              },
            ],
          },
        ],
        returns: {
          kind: 12,
        },
      },
    });
  });
  it('class-prop', () => {
    const results = parseFiles([path.resolve(__dirname, 'class-prop.ts')]);
    expect(results).toEqual({
      Boz: {
        name: 'Boz',
        kind: 13,
        properties: [
          {
            name: 'fn',
            kind: 11,
            parameters: [
              {
                name: 'foo',
                kind: 13,
                type: 'Foo',
                properties: [
                  {
                    parent: 'Foo',
                    static: true,
                    readonly: true,
                    name: 'dummy',
                    kind: 3,
                    value: false,
                  },
                ],
              },
            ],
            returns: {
              kind: 12,
            },
          },
        ],
      },
    });
  });
  it('async-function', () => {
    const results = parseFiles([path.resolve(__dirname, 'async-function.ts')]);
    expect(results).toEqual({
      genMyClass: {
        async: true,
        name: 'genMyClass',
        kind: 11,
        returns: {
          description: 'Represents the completion of an asynchronous operation',
          kind: 14,
          name: 'Promise',
          type: 'PromiseConstructor',
        },
      },
    });
  });

  it('generic-parameter', () => {
    const results = parseFiles(
      [path.resolve(__dirname, 'generic-parameter.ts')],
      { collectHelpers: true },
    );
    expect(results).toEqual({
      genericFunction: {
        name: 'genericFunction',
        kind: 11,
        parameters: [
          {
            name: 'box',
            kind: 14,
            generics: [
              {
                name: 'Type',
              },
            ],
            type: 'GenericInterface',
            properties: [
              {
                parent: 'GenericInterface',
                name: 'm',
                type: 'T',
              },
            ],
          },
          {
            name: 'newContents',
            type: 'Type',
          },
        ],
        returns: {
          kind: 14,
          generics: [
            {
              name: 'T',
            },
          ],
          name: 'GenericInterface',
          type: 'GenericInterface',
          properties: [
            {
              name: 'm',
              type: 'T',
            },
          ],
        },
        types: [
          {
            name: 'Type',
          },
        ],
      },
      __helpers: {
        GenericInterface: {
          name: 'GenericInterface',
          kind: 14,
          generics: [
            {
              name: 'T',
            },
          ],
          properties: [
            {
              name: 'm',
              type: 'T',
            },
          ],
        },
      },
    });
  });

  it('infer-return', () => {
    const results = parseFiles([path.resolve(__dirname, 'infer-return.ts')]);
    expect(results).toEqual({
      fn: {
        name: 'fn',
        kind: 11,
        parameters: [
          {
            name: 'in1',
            kind: 2,
          },
          {
            name: 'in2',
            kind: 2,
          },
        ],
        returns: {
          kind: 2,
        },
      },
    });
  });
  it('react fc', () => {
    const results = parseFiles([path.resolve(__dirname, 'react-fc.ts')]);
    expect(results).toMatchSnapshot();
  });
  it('function properties', () => {
    const results = parseFiles([path.resolve(__dirname, 'function-props.ts')]);

    expect(results).toEqual({
      fn: {
        name: 'fn',
        kind: 11,
        properties: [
          {
            kind: 1,
            description: 'custom property for a function',
            name: 'customProp',
            value: 'my custom prop',
          },
        ],
        returns: {
          kind: 12,
        },
      },
    });
  });

  it('object parameter', () => {
    const results = parseFiles([
      path.resolve(__dirname, 'object-parameter.ts'),
    ]);

    expect(results).toEqual({
      printCoord: {
        parameters: [
          {
            name: 'pt',
            description: 'object parameter',
            kind: 15,
            properties: [
              {
                description: 'x coordinate',
                kind: 2,
                name: 'x',
              },
              {
                description: 'optional y coordinate',
                kind: 2,
                optional: true,
                name: 'y',
              },
            ],
          },
        ],
        description: 'print coordinates',
        name: 'printCoord',
        kind: 11,
        returns: {
          kind: 12,
        },
      },
    });
  });

  it('tuple parameters', () => {
    const results = parseFiles([path.resolve(__dirname, 'tuple-parameter.ts')]);
    expect(results).toEqual({
      distanceFromOrigin: {
        name: 'distanceFromOrigin',
        kind: 11,
        parameters: [
          {
            kind: 6,
            properties: [
              {
                kind: 2,
              },
              {
                kind: 2,
              },
            ],
          },
        ],
        returns: {
          kind: 2,
        },
      },
    });
  });
  it('spread tuple parameters', () => {
    const results = parseFiles([
      path.resolve(__dirname, 'spread-tuple-parameter.ts'),
    ]);
    expect(results).toEqual({
      spreadTupleFunction: {
        name: 'spreadTupleFunction',
        kind: 11,
        parameters: [
          {
            name: 'args',
            kind: 6,
            properties: [
              {
                kind: 1,
              },
              {
                kind: 2,
              },
              {
                kind: 7,
                type: {
                  kind: 16,
                  properties: [
                    {
                      kind: 3,
                    },
                  ],
                },
              },
            ],
          },
        ],
        returns: {
          kind: 12,
        },
      },
    });
  });

  it('union parameter', () => {
    const results = parseFiles([path.resolve(__dirname, 'union-parameter.ts')]);
    expect(results).toEqual({
      printId: {
        name: 'printId',
        kind: 11,
        parameters: [
          {
            name: 'id',
            kind: 4,
            properties: [
              {
                kind: 2,
              },
              {
                kind: 1,
              },
            ],
          },
        ],
        returns: {
          kind: 12,
        },
      },
    });
  });

  it('jsdoc parameter', () => {
    const results = parseFiles([path.resolve(__dirname, 'jsdoc-parameter.ts')]);
    expect(results).toEqual({
      greet: {
        parameters: [
          {
            name: 'name',
            description: 'string type parameters',
            kind: 1,
          },
        ],
        description: 'greeting function',
        name: 'greet',
        kind: 11,
        returns: {
          kind: 12,
        },
      },
    });
  });
  it('arrow fn jsdoc', () => {
    const results = parseFiles([path.resolve(__dirname, 'arrow-function.ts')]);
    expect(results).toEqual({
      arrowGreet: {
        description: 'arrow greeting function',
        name: 'arrowGreet',
        kind: 11,
        parameters: [
          {
            description: 'name parameter inline description',
            kind: 1,
            name: 'name',
          },
        ],
        returns: {
          kind: 12,
        },
      },
    });
  });
});
