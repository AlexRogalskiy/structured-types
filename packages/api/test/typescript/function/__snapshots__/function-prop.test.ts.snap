// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`function react fc 1`] = `
Object {
  "TypedInitializedFunction": Object {
    "generics": Array [
      Object {
        "name": "P",
      },
    ],
    "kind": 11,
    "name": "TypedInitializedFunction",
    "parameters": Array [
      Object {
        "generics": Array [
          Object {
            "name": "P",
          },
        ],
        "kind": 15,
        "name": "props",
        "properties": Array [
          Object {
            "kind": 1,
            "name": "m",
            "optional": true,
            "parent": "Props",
          },
          Object {
            "kind": 4,
            "name": "children",
            "optional": true,
            "parent": "PropsWithChildren",
            "properties": Array [
              Object {
                "kind": 4,
                "name": "ReactNode",
                "properties": Array [
                  Object {
                    "kind": 4,
                    "name": "ReactChild",
                    "properties": Array [
                      Object {
                        "generics": Array [
                          Object {
                            "name": "P",
                          },
                          Object {
                            "name": "T",
                          },
                        ],
                        "kind": 14,
                        "name": "ReactElement",
                        "properties": Array [
                          Object {
                            "kind": 15,
                            "name": "type",
                            "type": "T",
                          },
                          Object {
                            "kind": 15,
                            "name": "props",
                            "type": "P",
                          },
                          Object {
                            "kind": 4,
                            "name": "key",
                            "properties": Array [
                              Object {
                                "kind": 4,
                                "name": "Key",
                                "properties": Array [
                                  Object {
                                    "kind": 1,
                                  },
                                  Object {
                                    "kind": 2,
                                  },
                                ],
                              },
                              Object {
                                "kind": 10,
                              },
                            ],
                          },
                        ],
                      },
                      Object {
                        "kind": 4,
                        "name": "ReactText",
                        "properties": Array [
                          Object {
                            "kind": 1,
                          },
                          Object {
                            "kind": 2,
                          },
                        ],
                      },
                    ],
                  },
                  Object {
                    "kind": 4,
                    "name": "ReactFragment",
                    "properties": Array [
                      Object {
                        "kind": 15,
                      },
                      Object {
                        "kind": 14,
                        "name": "ReactNodeArray",
                        "properties": Array [
                          Object {
                            "index": Object {
                              "kind": 2,
                              "name": "n",
                            },
                            "kind": 20,
                            "type": Object {
                              "kind": 15,
                              "name": "T",
                            },
                          },
                        ],
                      },
                    ],
                  },
                  Object {
                    "kind": 14,
                    "name": "ReactPortal",
                    "properties": Array [
                      Object {
                        "kind": 4,
                        "name": "key",
                        "properties": Array [
                          Object {
                            "kind": 4,
                            "name": "Key",
                            "properties": Array [
                              Object {
                                "kind": 1,
                              },
                              Object {
                                "kind": 2,
                              },
                            ],
                          },
                          Object {
                            "kind": 10,
                          },
                        ],
                      },
                      Object {
                        "kind": 15,
                        "name": "children",
                        "type": "ReactNode",
                      },
                      Object {
                        "kind": 15,
                        "name": "type",
                        "parent": "ReactElement",
                        "type": "T",
                      },
                      Object {
                        "kind": 15,
                        "name": "props",
                        "parent": "ReactElement",
                        "type": "P",
                      },
                    ],
                  },
                  Object {
                    "kind": 3,
                  },
                  Object {
                    "kind": 10,
                  },
                  Object {
                    "kind": 8,
                  },
                ],
              },
              Object {
                "kind": 8,
              },
            ],
          },
        ],
      },
    ],
    "properties": Array [
      Object {
        "kind": 4,
        "name": "propTypes",
        "optional": true,
        "parent": "FunctionComponent",
        "properties": Array [
          Object {
            "generics": Array [
              Object {
                "name": "T",
              },
            ],
            "kind": 15,
            "name": "WeakValidationMap",
          },
          Object {
            "kind": 8,
          },
        ],
      },
      Object {
        "kind": 4,
        "name": "contextTypes",
        "optional": true,
        "parent": "FunctionComponent",
        "properties": Array [
          Object {
            "generics": Array [
              Object {
                "name": "T",
              },
            ],
            "kind": 15,
            "name": "ValidationMap",
          },
          Object {
            "kind": 8,
          },
        ],
      },
      Object {
        "kind": 4,
        "name": "defaultProps",
        "optional": true,
        "parent": "FunctionComponent",
        "properties": Array [
          Object {
            "description": "Make all properties in T optional",
            "generics": Array [
              Object {
                "name": "T",
              },
            ],
            "kind": 15,
            "name": "Partial",
          },
          Object {
            "kind": 8,
          },
        ],
      },
      Object {
        "kind": 4,
        "name": "displayName",
        "optional": true,
        "parent": "FunctionComponent",
        "properties": Array [
          Object {
            "kind": 1,
          },
          Object {
            "kind": 8,
          },
        ],
      },
    ],
  },
}
`;
